Add support for cross module inlining and cross module generic compilation to Crossgen2

davidwrighton opened the pull request:
- Refactor Module into ModuleBase and Module
  - The goal is to have allow a subset version of Module which can only hold refs, this is to be used by the manifest module in an R2R image to allow for version resilient cross module references.
  - Update handling of ModuleBase so that its used everywhere that tokens are parsed from R2R
  - Remove ENCODE_MODULE_ID_FOR_STATICS and ENCODE_ACTIVE_DEPENDENCY
    - These were only used for NGEN, and conflict with easy impelmentation for the ModuleBase concept
  - Remove locking in ENCODE_STRING_HANDLE processing, and tweak comments. Comments applied to the removed ngen based code, and the lock was only necessary for the old ngen thing.
  - Adjust ComputeLoaderModuleWorker for locating loader module
    - Follow comment more accurately, to avoid putting every generic into its definition module. This will make R2R function lookup able to find compiled instantiations in some cases. This may be what we want long term, it may not.
  - Remove MemberRefToDesc map and replace with LookupMap like the other token types. We no longer make use of the hot table, so this is more efficient
    - Also reduces complexity of implementation of ModuleBase

- Build fixup to describe a single method as a standalone blob of data
  - There are parallel implementations in Crossgen2 and in the runtime
  - They produce binary identical output
  - Basic R2RDump support for new fixup

- Adjust module indices used within the R2R format to support a module index which refers to the R2R manifest metadata. This requires bumping the R2R version to 6.2
  - Add a module index between the set of assembly refs in the index 0 module and the set of assembly refs in the R2R manifest metadata

- Adjust compilation dependency rules to include a few critical AsyncStateMachineBox methods

- Remove PEImage handling of native metadata which was duplicative

- Do not enable any more devirtualization than was already in use, even in the cross module compilation scenario. In particular, do not enable devirtualization of methods where the decl method isn't within the version bubble, even if the decl method could be represented with a cross module reference token. (This could be fixed, but is out of scope for this initial investigation)

Make the compilation deterministic in this new model, even though we are generating new tokens on demand
  - Implement this by detecting when we need new tokens during a compile, and recompiling with new tokens when necessary
  - This may result in compiling the same code as much as twice

Compile the right set of methods with cross module inlining enabled
- Add support for compiling the called virtual methods on generic types
  - This catches the List<T> and Dictionary<TKey,TValue> scenarios

- Add command line switches to enable/disable the new behavior

trylek reviewed (commented):
First batch of comments towards David's change, still going over the code.

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunStandalongMethodMetadata.cs:
```diff
@@ -0,0 +1,229 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Definitely a typo...

davidwrighton commented on src/coreclr/tools/Common/TypeSystem/Mutable/MutableModule.cs:
```diff
@@ -0,0 +1,315 @@
+                    // No support for non-system module references yet
+                }
+
+                _moduleRefStrings.Add("System.Private.CoreLib");
```
The dictionary check in line 27 prevents the repeated addition to the list. Effectively, this is currently a bit of a stub, which will be filled out in a future PR. In addition, we explicitly don't want a HashSet here, as the list represents an ordering. The dictionary in line 27 provides the control over excessive module ref string generation.

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilationModuleGroupBase.cs:
```diff
@@ -337,12 +350,107 @@ public sealed override bool CanInline(MethodDesc callerMethod, MethodDesc callee
+                }
+            }
+
+            return true;
```
The difference between the two is that one represents the set of things that can be inlined which is strictly greater than the set of methods to actually compile.

Think of the method `Span<int>.get_Item(int index)`. This is a method which is fine to inline, but there is no good reason to allow it to be compiled as a separate method. In addition the inlineable check will allow for non-generics to be compiled. As you say though, the current implementations are very similar. I do expect though, over time that these implementation may change more radically over time.

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunStandalongMethodMetadata.cs:
```diff
@@ -0,0 +1,229 @@
+                    var exceptionRegion = exceptionRegions[i];
+
+                    _exceptionRegions[i] = new ILExceptionRegion(
+                        (ILExceptionRegionKind)exceptionRegion.Kind, // assumes that ILExceptionRegionKind and ExceptionRegionKind enums are in sync
```
Eh, there is quite a lot of dependence on these being identical, also there is the same comment in the EcmaMethodILProvider as well. This will be fine.

trylek commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/IL/ReadyToRunILProvider.cs:
```diff
@@ -104,7 +155,11 @@ public override MethodIL GetMethodIL(MethodDesc method)
-                MethodIL methodIL = EcmaMethodIL.Create(ecmaMethod);
+                if (!_manifestModuleWrappedMethods.TryGetValue(ecmaMethod, out var methodIL))
+                {
+                    methodIL = EcmaMethodIL.Create(ecmaMethod);
```
Should we call <code>CreateCrossModuleInlineableTokensForILBody</code> or add the created <code>methodIL</code> to the <code>_manifestModuleWrappedMethods</code> cache here?

davidwrighton replied:
No, that would cause determinism bugs in the compiler. I'll add a comment.

trylek commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -1985,7 +2064,7 @@ private void getCallInfo(ref CORINFO_RESOLVED_TOKEN pResolvedToken, CORINFO_RESO
                         {
                             // READYTORUN: FUTURE: Direct calls if possible
-                            pResult->codePointerOrStubLookup.constLookup = CreateConstLookupToSymbol(
+                             pResult->codePointerOrStubLookup.constLookup = CreateConstLookupToSymbol(
```
Nit - extra space breaking code alignment in the block.

trylek commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -2636,13 +2715,88 @@ private void setEHinfo(uint EHnumber, ref CORINFO_EH_CLAUSE clause)
+
+                if (_precodeFixups != null)
+                {
+                    previouslyStashedFixups = previouslyStashedFixups ?? new List<ISymbolNode>();
```
Non-blocking nit - when <code>previouslyStashedFixups</code> is null, you could directly assign <code>_precodeFixups</code> to it and avoid copying. Not sure how prevalent this is in practice.

trylek commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -2636,13 +2715,88 @@ private void setEHinfo(uint EHnumber, ref CORINFO_EH_CLAUSE clause)
+                HashSet<MethodDesc> previouslyStashedInlinees = _stashedInlinedMethods.Pop();
+                if (_inlinedMethods != null)
+                {
+                    previouslyStashedInlinees = previouslyStashedInlinees ?? new HashSet<MethodDesc>();
```
dtto

trylek commented on src/coreclr/vm/clsload.cpp:
```diff
@@ -91,33 +118,42 @@ PTR_Module ClassLoader::ComputeLoaderModuleWorker(
             pLoaderModule = pModule;
     }
 
+    if ((pLoaderModule == NULL) && pDefinitionModule)
```
Nit - the check for <code>pDefinitionModule</code> isn't really necessary here, I guess it's just a matter of your aesthetic feeling ;-).

trylek commented on src/coreclr/vm/inlinetracking.cpp:
```diff
@@ -424,7 +424,235 @@ Module* PersistentInlineTrackingMapR2R2::GetModuleByIndex(DWORD index)
+    return TRUE;
+}
+
+// Read a Module/MethodDef token pair from the parser. Optionally check to see if it matches expectedToken. 
```
Looks like a copy-paste comment unrelated to actual behavior of this method.

trylek commented on src/coreclr/vm/inlinetracking.cpp:
```diff
@@ -424,7 +424,235 @@ Module* PersistentInlineTrackingMapR2R2::GetModuleByIndex(DWORD index)
+
+    // This "black magic spell" has in fact nothing to do with GenericInstantiationCompare per se, but just sets a thread flag
+    // that later activates more thorough search inside Module::GetAssemblyIfLoaded, which is indirectly called from GetModuleFromIndexIfLoaded.
+    // This is useful when ngen image was compiler against a different assembly version than the one loaded now.
```
Nit - typo (compiler -> compiled). Also, do we still want to refer to R2R images as "ngen"?

trylek commented on src/coreclr/vm/method.cpp:
```diff
@@ -4099,3 +4099,694 @@ void MethodDesc::PrepareForUseAsAFunctionPointer()
+
+class AlternateBlockHelper
+{
+//    COR_ILMETHOD* pHeader;
```
Is there any value in keeping this code comment?

trylek commented on src/coreclr/vm/method.cpp:
```diff
@@ -4099,3 +4099,694 @@ void MethodDesc::PrepareForUseAsAFunctionPointer()
+public:
+
+    AlternateBlockHelper(MethodDesc *pMD, SArray<uint32_t> *pTypeRefTokenStreamInput) :
+//        pHeader(pMD->GetILHeader(TRUE)),
```
dtto

trylek commented on src/coreclr/vm/method.cpp:
```diff
@@ -4099,3 +4099,694 @@ void MethodDesc::PrepareForUseAsAFunctionPointer()
+        // ModuleRef is encoded with low bits 010
+        // MethodDef is encoded with low bit 011
+        // TypeSpec is encoded with low bits 100
+        for (uint32_t codedIndexType = 0; codedIndexType <= 7; codedIndexType++)
```
Hmm, this looks pretty inefficient, do we know it's used seldom?

trylek commented on src/coreclr/vm/method.cpp:
```diff
@@ -4099,3 +4099,694 @@ void MethodDesc::PrepareForUseAsAFunctionPointer()
+        ThrowHR(COR_E_BADIMAGEFORMAT);
+    }
+
+    uint32_t GetAlternateToken(uint32_t inputToken)
```
It might be useful to mention a reference to the equivalent C# code in a code comment somewhere around here.

trylek commented on src/coreclr/vm/readytoruninfo.cpp:
```diff
@@ -626,13 +627,31 @@ ReadyToRunInfo::ReadyToRunInfo(Module * pModule, LoaderAllocator* pLoaderAllocat
         m_pComposite = &m_component;
         m_isComponentAssembly = false;
+        IMDInternalImport *pNativeMDImport;
+        IMAGE_DATA_DIRECTORY * pNativeMetadatSection = m_pComposite->FindSection(ReadyToRunSectionType::ManifestMetadata);
```
Nit - typo, should be pNativeMetadat**a**Section I guess.

trylek commented on src/coreclr/vm/readytoruninfo.cpp:
```diff
@@ -706,6 +725,17 @@ ReadyToRunInfo::ReadyToRunInfo(Module * pModule, LoaderAllocator* pLoaderAllocat
         m_entryPointToMethodDescMap.Init(TRUE, &lock);
     }
 
+    if (IsImageVersionAtLeast(6, 2))
```
Why do we need to check the R2R format version? I was under the impression that we reject any older R2R file upfront so we don't even get here.

trylek commented on src/coreclr/vm/readytoruninfo.cpp:
```diff
@@ -862,10 +935,18 @@ bool ReadyToRunInfo::GetPgoInstrumentationData(MethodDesc * pMD, BYTE** pAllocat
     return false;
 }
 
+//#define LOG_R2R_ENTRYPOINT
```
Is this comment still useful?

trylek reviewed (commented):
I think that by now I have actually seen all the changes in the PR, and while I cannot mentally emulate all of them at once, all I've seen in general makes sense and seems reasonable to me.

davidwrighton committed:
Add support for cross module inlining and cross module generic compilation to Crossgen2
- Refactor Module into ModuleBase and Module
  - The goal is to have allow a subset version of Module which can only hold refs, this is to be used by the manifest module in an R2R image to allow for version resilient cross module references.
  - Update handling of ModuleBase so that its used everywhere that tokens are parsed from R2R
  - Remove ENCODE_MODULE_ID_FOR_STATICS and ENCODE_ACTIVE_DEPENDENCY
    - These were only used for NGEN, and conflict with easy impelmentation for the ModuleBase concept
  - Remove locking in ENCODE_STRING_HANDLE processing, and tweak comments. Comments applied to the removed ngen based code, and the lock was only necessary for the old ngen thing.
  - Adjust ComputeLoaderModuleWorker for locating loader module
    - Follow comment more accurately, to avoid putting every generic into its definition module. This will make R2R function lookup able to find compiled instantiations in some cases. This may be what we want long term, it may not.
  - Remove MemberRefToDesc map and replace with LookupMap like the other token types. We no longer make use of the hot table, so this is more efficient
    - Also reduces complexity of implementation of ModuleBase

- Build fixup to describe a single method as a standalone blob of data
  - There are parallel implementations in Crossgen2 and in the runtime
  - They produce binary identical output
  - Basic R2RDump support for new fixup

- Adjust module indices used within the R2R format to support a module index which refers to the R2R manifest metadata. This requires bumping the R2R version to 6.2
  - Add a module index between the set of assembly refs in the index 0 module and the set of assembly refs in the R2R manifest metadata

- Adjust compilation dependency rules to include a few critical AsyncStateMachineBox methods

- Remove PEImage handling of native metadata which was duplicative

- Do not enable any more devirtualization than was already in use, even in the cross module compilation scenario. In particular, do not enable devirtualization of methods where the decl method isn't within the version bubble, even if the decl method could be represented with a cross-module reference token. (This could be fixed, but is out of scope for this initial investigation)

Make the compilation deterministic in this new model, even though we are generating new tokens on demand
  - Implement this by detecting when we need new tokens during a compile, and recompiling with new tokens when necessary
  - This may result in compiling the same code as much as twice

Compile the right set of methods with cross module inlining enabled
- Add support for compiling the called virtual methods on generic types
  - This catches the List<T> and Dictionary<TKey,TValue> scenarios

- Add command line switches to enable/disable the new behavior

NOTE: Module::IsInSameVersionBubble doesn't actually behave as expected. Currently we're generating a native manifest metadata in all circumstances, which will cause all modules to be treated as having InputBubble semantics. However, fortunately, we don't actually do anything like that at runtime. The only use of IsInSameVersionBubble is used to control R2R function lookup, which will not cause a problem.

davidwrighton committed:
Temporary workaround

davidwrighton committed:
Disable cross module pinvoke inlining
- We don't model the details necessary to prevent this from possibly being a problem at this time

davidwrighton committed:
Make the loader module algorithm more similar to calculations in crossgen2

davidwrighton committed:
Fix Mono.Linker.Tests

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/IL/ReadyToRunILProvider.cs:
```diff
@@ -130,5 +185,93 @@ public override MethodIL GetMethodIL(MethodDesc method)
+                    }
+                    if (!newToken.HasValue)
+                    {
+                        throw new Exception("Nope2");
```
I've replaced it with a pattern where if this failure occurs, then the method is silently dropped. This will be needed for future work where we expand the set of cross module compilable methods beyond just that of System.Private.CoreLib

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.Reflection.ReadyToRun/ReadyToRunSignature.cs:
```diff
@@ -1439,6 +1456,20 @@ private ReadyToRunSignature ParseSignature(ReadyToRunFixupKind fixupType, String
+                    SkipBytes(ilBodyByteBlobSize);
+                    uint types = ReadUInt();
+                    for (uint i = 0; i < types; i++)
+                        ParseType();
```
This call to ParseType will print the type, but yes, the rest of the data is hidden. I struggled with this one, as the output is ... likely to be extremely large for relatively low value. My approach instead was to use the verify il body stuff to ensure that the runtime and crossgen2 agreed on the format, and just to skip actual emission.

trylek replied:
Sure, that sounds reasonable for now, if we hit a particularly tricky bug involving this data in the future, we can always extend R2RDump by adding more detailed logging of the fixup.

davidwrighton committed:
Address code review feedback, and fix naming for native side of ReadyToRunStandaloneMethodMetadata

davidwrighton committed:
Update the R2R docs

davidwrighton committed:
Fix issues found in testing

davidwrighton committed:
Fix markdownlint

davidwrighton committed:
Handle RawArrayData use in intrinsic in better way

davidwrighton committed:
Adjust command line switches for new behavior to be disabled, and explicitly declared to be experimental

davidwrighton committed:
Fix the build

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/IL/ILTokenReplacer.cs:
```diff
@@ -0,0 +1,163 @@
+            int currentOffset = 0;
+            for (; currentOffset < ilStream.Length;)
+            {
+                ILOpcode opCode = (ILOpcode)ReadILByte();
```
I think using ILReader for parsing would make this a lot shorter.

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/IL/NativeAotILProvider.cs:
```diff
@@ -14,6 +16,13 @@ namespace Internal.IL
     {
+        private readonly CompilationModuleGroup _compilationGroup;
+
+        public NativeAotILProvider(CompilationModuleGroup group)
```
Could you instead `#ifdef READYTORUN` the InterlockedIntrinsics to not take CompilationModuleGroup outside READYTORUN?

ILProvider taking a dependency on ILCompiler stuff is odd and confusing. Nothing under Common\TypeSystem should reference ILCompiler namespace as a test for layering violations.

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/IL/ILProvider.cs:
```diff
@@ -11,6 +11,12 @@ namespace Internal.IL
     /// </summary>
     public abstract class ILProvider
     {
+        private int _version = 0;
+        protected void IncrementVersion()
+        {
+            _version++;
+        }
+        public int Version => _version;
         public abstract MethodIL GetMethodIL(MethodDesc method);
```
Could you move this to `ReadyToRunILProvider` and retype a couple things to fix build breaks? There will be one place where hard cast is necessary in `UseILProvider(ILProvider ilProvider)`. I think that's fine.

MichalStrehovsky commented on src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -1205,6 +1222,11 @@ private void getMethodSig(CORINFO_METHOD_STRUCT_* ftn, CORINFO_SIG_INFO* sig, CO
         private bool getMethodInfo(CORINFO_METHOD_STRUCT_* ftn, CORINFO_METHOD_INFO* info)
         {
             MethodDesc method = HandleToObject(ftn);
+            // Add an early CanInline check to see if referring to the IL of the target methods is
+            // permitted from within this MethodBeingCompiled, the full CanInline check will be performed
+            // later.
+            if (!_compilation.CanInline(MethodBeingCompiled, method))
+                return false;
```
```suggestion
#if READYTORUN
            // Add an early CanInline check to see if referring to the IL of the target methods is
            // permitted from within this MethodBeingCompiled, the full CanInline check will be performed
            // later.
            if (!_compilation.CanInline(MethodBeingCompiled, method))
                return false;
#endif
```

davidwrighton committed:
Fix break for R2R binaries without cross module inlining enabled

davidwrighton committed:
Move shash to heap from global, just to see what it does

davidwrighton committed:
Address Michal's feedback

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into cross_module_r2r

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into cross_module_r2r

trylek commented on docs/design/coreclr/botr/readytorun-format.md:
```diff
@@ -578,6 +608,33 @@ Number of assemblies stored in the manifest metadata is equal to the number of M
+## ReadyToRunSectionType.CrossModuleInlineInfo (v6.2+)
+The inlining information section captures what methods got inlined into other methods. It consists of a single _Native Format Hashtable_ (described below).
+
+The entries in the hashtable are lists of inliners for each inlinee. One entry in the hashtable corresponds to one inlinee. The hashtable is hashed the version resilient hashcode of the uninstantiated methoddef inlinee.
```
Wording nit - should it be "The hashtable is hashed *with* the version resilient hashcode"? I'm not sure about the preposition but a preposition seems appropriate to me in this position.

davidwrighton replied:
Fixed

trylek commented on docs/design/coreclr/botr/readytorun-format.md:
```diff
@@ -578,6 +608,33 @@ Number of assemblies stored in the manifest metadata is equal to the number of M
+
+The entries in the hashtable are lists of inliners for each inlinee. One entry in the hashtable corresponds to one inlinee. The hashtable is hashed the version resilient hashcode of the uninstantiated methoddef inlinee.
+
+The entry of the hashtable is a counted sequence of compressed unsigned integers which begins with an InlineeIndex which has flags which then defines how a sequence of inliners may be read from the file.
```
Nit - "flags ... which then define" - or would it be perhaps easier for a future reader to put it like "flags defining how a sequence..."?

davidwrighton replied:
Reworded to "The entry of the hashtable is a counted sequence of compressed unsigned integers which begins with an InlineeIndex which combines a 30 bit index with 2 bits of flags which how the sequence of inliners shall be parsed and what table is to be indexed into to find the inlinee."

trylek commented on docs/design/coreclr/botr/readytorun-format.md:
```diff
@@ -578,6 +608,33 @@ Number of assemblies stored in the manifest metadata is equal to the number of M
+
+* InlineeIndex
+  * Index with 2 flags field in lowest 2 bits to define the inlinee
+    - If flags & 1 == 0 then index is a MethodDef RID, and if the module is a composite image, a module index of the method follows
```
Nit - assuming this is supposed to be a C++ / C# pseudo-code sequence, I believe the bitwise and operator would need to be parenthesized. Not sure if it's worth addressing.

davidwrighton replied:
Done.

trylek commented on docs/design/coreclr/botr/readytorun-format.md:
```diff
@@ -578,6 +608,33 @@ Number of assemblies stored in the manifest metadata is equal to the number of M
+      - the sequence of delta encoded indices into the first import section with a type of READYTORUN_IMPORT_SECTION_TYPE_ILBODYFIXUPS
+      - Inliner RID deltas - See definition below
+
+* Inliner RID deltas (for multi-module version bubble images (specified by the module having the READYTORUN_FLAG_MULTIMODULE_VERSION_BUBBLE flag set)
```
Nit - two opening parentheses with just one closing, I would remove the second opening parenthesis.

trylek commented on docs/design/coreclr/botr/readytorun-format.md:
```diff
@@ -124,12 +125,15 @@ struct READYTORUN_CORE_HEADER
-| READYTORUN_FLAG_COMPONENT               | 0x00000008 | This is a component assembly of a composite R2R image
+| Flag                                       |      Value | Description
+|:-------------------------------------------|-----------:|:-----------
+| READYTORUN_FLAG_PLATFORM_NEUTRAL_SOURCE    | 0x00000001 | Set if the original IL image was platform neutral. The platform neutrality is part of assembly name. This flag can be used to reconstruct the full original assembly name.
```
Thanks for fixing the obsolete values in this table!

trylek commented on src/coreclr/tools/Common/TypeSystem/Mutable/MutableModule.Sorting.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    partial class MutableModule
+    {
+        // This isn't deterministic, but it is functional.
```
Am I right to understand this actually doesn't violate multithreaded Crossgen2 determinism because in practice there's always at most one mutable module per compilation?

davidwrighton replied:
Yes. I've updated the comment.

trylek commented on src/coreclr/vm/prestub.cpp:
```diff
@@ -503,21 +503,33 @@ PCODE MethodDesc::GetPrecompiledR2RCode(PrepareCodeConfig* pConfig)
-        pModule = SystemDomain::System()->DefaultDomain()->GetRootAssembly()->GetModule();
-        _ASSERT(pModule != NULL);
+        Module* pDefiningModule = GetModule();
+        // Lookup in the defining module of the generic (which is where in inputbubble scenarios
```
Wording nit - I think that the word "that" at line 517 is superfluous and there's a missing closing parenthesis.

trylek commented on src/coreclr/vm/readytorunstandalonemethodmetadata.cpp:
```diff
@@ -0,0 +1,694 @@
+                        StandaloneSigTranslator sigTranslator(&memberRefSigParse, &blob, this);
+                        sigTranslator.ParseMemberRefSignature();
+                        ULONG strLen = (ULONG)strlen(name); // Cast to ULONG is safe, as the data is held in a PE file
+                        blob.AppendData((ULONG)strlen(name));
```
Nit - do we need to calculate <code>strlen</code> a second time considering we have evaluated it at line 294?

trylek commented on src/coreclr/vm/readytorunstandalonemethodmetadata.cpp:
```diff
@@ -0,0 +1,694 @@
+                        StandaloneSigTranslator sigTranslator(&methodDefSigParse, &blob, this);
+                        sigTranslator.ParseMethodSignature();
+                        ULONG strLen = (ULONG)strlen(name); // Cast to ULONG is safe, as the data is held in a PE file
+                        blob.AppendData((ULONG)strlen(name));
```
dtto

trylek commented on src/coreclr/vm/readytorunstandalonemethodmetadata.cpp:
```diff
@@ -0,0 +1,694 @@
+                        StandaloneSigTranslator sigTranslator(&fieldDefSigParse, &blob, this);
+                        sigTranslator.ParseFieldSignature();
+                        ULONG strLen = (ULONG)strlen(name); // Cast to ULONG is safe, as the data is held in a PE file
+                        blob.AppendData((ULONG)strlen(name));
```
dtto

trylek reviewed (approved):
Looks great to me overall, I have just some minor nitpicky comments and questions. Thanks a lot for pulling this off David, this is a beast of a change!

trylek commented on src/coreclr/vm/versionresilienthashcode.cpp:
```diff
@@ -71,6 +71,40 @@ bool GetVersionResilientTypeHashCode(IMDInternalImport *pMDImport, mdExportedTyp
+
+    LPCUTF8 szName;
+
+    if (FAILED(pMDImport->GetNameOfMethodDef(token, &szName)))
```
Thanks for clarifying, I thought it would be the case but I just wanted to double-check.

davidwrighton committed:
Address feedback final

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Mutable/MutableModule.Sorting.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    partial class MutableModule
+    {
+        // This isn't deterministic, but it is functional. At this time, since only 1 MutableModule is contained in a build, it will be deterministic as it will always return 0.
```
I already felt that maybe this shouldn't in the common type system, but Tomas's comment brought me over the fence.

Determinism and thread safety are the core invariants in the type system.

Could all of this just go to ILCompiler.ReadyToRun? It doesn't really look reusable.

davidwrighton committed:
Move MutableModule around per Michal's request

davidwrighton committed:
Fix error where the load in LookupMap can actually be torn in the presence of multithreading.

davidwrighton committed:
Update R2R version bump to 6.3 as this missed checking in ahead of some work by Jakob

davidwrighton committed:
Merge branch 'main' into cross_module_r2r
